lib/src/perf_benchmark_base.dart

```diff
@@ -0,0 +1,146 @@
+import 'dart:convert';
+import 'dart:io';
+
+import 'benchmark_base.dart';
+import 'score_emitter.dart';
+
+class PerfBenchmarkBase extends BenchmarkBase {
+  PerfBenchmarkBase(super.name, {super.emitter = const PrintEmitter()});
+
+  late final Directory fifoDir;
+  late final String perfControlFifo;
+  late final RandomAccessFile openedFifo;
+  late final String perfControlAck;
+  late final RandomAccessFile openedAck;
+  late final Process perfProcess;
+
+  Future<void> _createFifos() async {
+    fifoDir = await Directory.systemTemp.createTemp('fifo');
+    perfControlFifo = '${fifoDir.path}/perf_control_fifo';
+    perfControlAck = '${fifoDir.path}/perf_control_ack';
+
+    try {
+      final fifoResult = await Process.run('mkfifo', [perfControlFifo]);
+      if (fifoResult.exitCode != 0) {
+        throw ProcessException('mkfifo', [perfControlFifo],
+            'Cannot create fifo: ${fifoResult.stderr}', fifoResult.exitCode);
+      }
+      final ackResult = await Process.run('mkfifo', [perfControlAck]);
+      if (ackResult.exitCode != 0) {
+        throw ProcessException('mkfifo', [perfControlAck],
+            'Cannot create fifo: ${ackResult.stderr}', ackResult.exitCode);
+      }
+    } catch (e) {
+      await fifoDir.delete(recursive: true);
+      rethrow;
+    }
+  }
+
+  Future<void> _startPerfStat() async {
+    await _createFifos();
+    try {
+      perfProcess = await Process.start('perf', [
+        'stat',
+        '--delay',
+        '-1',
+        '--control',
+        'fifo:$perfControlFifo,$perfControlAck',
+        '-j',
+        '-p',
+        '$pid'
+      ]);
+      //await Future<void>.delayed(const Duration(seconds: 2));
+
+      openedFifo = File(perfControlFifo).openSync(mode: FileMode.writeOnly);
+
+      openedAck = File(perfControlAck).openSync();
+      openedFifo.writeStringSync('enable\n');
+      _waitForAck();
+    } catch (e) {
+      await fifoDir.delete(recursive: true);
+      rethrow;
+    }
+  }
+
+  Future<void> _stopPerfStat(int totalIterations) async {
+    try {
+      openedFifo.writeStringSync('disable\n');
+      openedFifo.closeSync();
+      _waitForAck();
+      openedAck.closeSync();
+      perfProcess.kill(ProcessSignal.sigint);
+      final lines =
+          utf8.decoder.bind(perfProcess.stderr).transform(const LineSplitter());
+      // Exit code from perf is -2 when terminated with SIGINT.
+      final exitCode = await perfProcess.exitCode;
+      if (exitCode != 0 && exitCode != -2) {
+        throw ProcessException(
+            'perf',
+            [
+              'stat',
+              '--delay',
+              '-1',
+              '--control',
+              'fifo:$perfControlFifo,$perfControlAck',
+              '-j',
+              '-p',
+              '$pid'
+            ],
+            (await lines.toList()).join('\n'),
+            exitCode);
+      }
+      final events = [
+        await for (final line in lines)
+          if (line.startsWith('{"counter-value" : '))
```

Comment:
This seems brittle, can you decode all lines as json and see if they have the right key?

Reply:
I would have to catch the exception when it isn't valid JSON. I could just check for the string"counter-value", and if that string is present, decode as json and do the further check.

Reply:
Because this is in a literal, I cannot use try-catch to skip creating a object if the line does not decode as correct JSON.  So unless I rewrite the code to use a for loop and try-catch to create a list of ojbects iteratively, I would just be replacing the fragility of breaking if perf changes its output format, to breaking if stderr contains the string "counter-value" in a line that isn't a JSON literal.

Done, changed to just a check that "counter-value" is in the line.